src/ndpp_init.F90

```diff
@@ -0,0 +1,604 @@
+module ndpp_initialize
+
+  use ace_header,   only: Nuclide, XsListing
+  use constants
+  use dict_header,  only: DictCharInt, DICT_NULL
+  use error,        only: fatal_error, warning
+  use global
+  use ndpp_header,  only: Ndpp
+  use ndpp_ops,     only: ndpp_read
+  use output,       only: write_message
+  use search
+  use string,       only: ends_with, to_lower, starts_with, to_str
+  use xml_interface
+
+  implicit none
+
+  ! ndpp_lib.xml preprocessed data listings and associated data.
+  type(XsListing), allocatable, target, save :: ndpp_listings(:)
+  type(DictCharInt), save                    :: ndpp_listing_dict
+
+contains
+
+!===============================================================================
+! READ_NDPP_LIB reads all the cross sections for the problem and stores them in
+! nuclides and sab_tables arrays
+!===============================================================================
+
+  subroutine read_ndpp_data()
+    type(Nuclide), pointer    :: nuc => null() ! Current working nuclide
+    type(SAlphaBeta), pointer :: sab => null() ! Current working SAB table
+    type(XsListing), pointer  :: ndpp_listing => null()
+    integer :: i_listing     ! index in ndpp_listings array
+    integer :: i_nuclide     ! index in nuclides
+    integer :: i_sab         ! index in sab_tables
+    integer :: scatt_type    ! Whether or not legendre or tabular data
+    logical :: get_scatt     ! Flag for whether or not to get scatt data
+    logical :: get_nuscatt   ! Flag for whether or not to get nuscatt data
+    logical :: get_chi_t     ! Flag for whether or not to get total chi data
+    logical :: get_chi_p     ! Flag for whether or not to get prompt chi data
+    logical :: get_chi_d     ! Flag for whether or not to get delayed chi data
+    integer :: scatt_order   ! Number of scatter moments requested in tallies
+    integer :: ndpp_groups   ! Number of groups in the NDPP library
+
+    ! allocate arrays for NDPP data storage
+    allocate(ndpp_nuc_data(n_nuclides_total))
+    allocate(ndpp_sab_data(n_sab_tables))
+
+    ! Read the ndpp_lib.xml file
+    call read_ndpp_xml(scatt_type, ndpp_groups)
+
+    ! Determine which data is required to be stored as well as the maximum
+    ! orders for scatt and nuscatt
+    call which_data(scatt_type, get_scatt, get_nuscatt, get_chi_t, get_chi_p, &
+                    get_chi_d, scatt_order)
+
+    ! Parse through each nuclide in the model and read in the corresponding
+    ! NDPP data.
+    ! During this we will make sure the temperatures of ACE and NDPP match,
+    ! as well as making sure the NDPP data exists at all.
+
+    do i_nuclide = 1, n_nuclides_total
+      nuc => nuclides(i_nuclide)
+      i_listing = ndpp_listing_dict % get_key(adjustl(trim(nuc % name)))
+      if (i_listing == DICT_NULL) then
```

Comment:
A more idiomatic way of doing this would be:
```fortran
if (.not. ndpp_listing_dict % has_key(adjustl(trim(nuc % name)))) then
```
Then you don't have to make `DICT_NULL` public.